astropy/coordinates/sky_coordinate.py

```diff
@@ -484,6 +487,94 @@ def transform_to(self, frame, merge_attributes=True):
             frame_kwargs.pop(attr)
         return self.__class__(new_coord, **frame_kwargs)
 
+    def apply_space_motion(self, new_obstime=None, dt=None):
+        """
+        Update the position of the source represented by this coordinate object
+        to a new specified time using the velocity and assuming linear motion.
+
+        This is sometimes referred to as an "epoch transformation."
+
+        Parameters
+        ----------
+        new_obstime : `~astropy.time.Time`, optional
+            The time at which to evolve the position to.
+        dt : `~astropy.units.Quantity`, `~astropy.time.TimeDelta`, optional
+            An amount of time to evolve the position of the source.
+        """
+
+        if (new_obstime is None and dt is None or
+                new_obstime is not None and dt is not None):
+            raise ValueError("You must specify one of `new_obstime` or `dt`, "
+                             "but not both.")
+
+        # Validate that we have velocity info
+        if 's' not in self.frame.data.differentials:
+            raise ValueError('SkyCoord requires velocity data to evolve the '
+                             'position.')
+
+        if 'obstime' in self.frame.frame_attributes:
+            raise NotImplementedError("Updating the coordinates in a frame "
+                                      "with explicit time dependence is "
+                                      "currently not supported. If you would "
+                                      "like this functionality, please open an "
+                                      "issue on github:\n"
+                                      "https://github.com/astropy/astropy")
+
+        if new_obstime is not None and self.obstime is None:
```

Comment:
This section creating `t1, t2` is very messy. How about calling them `old_obstime` and `new_obstime` and doing the following:
```
old_obstime = self.obstime
if new_obstime is None:
    if old_obstime is None:
           # MAGIC NUMBER: if the current SkyCoord object has no obstime,
           # assume J2000 to do the dt offset.
           old_obstime = Time('J2000')
    new_obstime = old_obstime + dt
elif old_obstime is None:
    # No default when a new time is passed in.
    raise(...)
```

Reply:
I don't really see how this is less messy...but I will change if you feel strongly (I don't!).

Reply:
Yes, I do feel it is less messy...

Reply:
While re-working this I realized it has to be a bit more messy than this because `t1` and `t2` are *not* quite the same as `new_obstime` and `old_obstime`.  So it's still a bit messy but I think now more correct.